sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,595 @@
+    public BoundedReader<String> createReader(PipelineOptions options) throws IOException {
+      return new BoundedElasticsearchReader(this);
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public Coder<String> getDefaultOutputCoder() {
+      return StringUtf8Coder.of();
+    }
+
+    private JsonObject getStats(boolean shardLevel) throws IOException {
+      RestClient restClient = spec.getConnectionConfiguration().createClient();
+      HashMap<String, String> params = new HashMap<>();
+      if (shardLevel) {
+        params.put("level", "shards");
+      }
+      String endpoint = String.format("/%s/_stats",
+                                      spec.getConnectionConfiguration
+                                          ().getIndex());
+      Response response =
+          restClient.performRequest("GET", endpoint, params, new BasicHeader("", ""));
+      restClient.close();
+      return parseResponse(response);
+
+    }
+
+  }
+
+  private static class BoundedElasticsearchReader extends BoundedSource.BoundedReader<String> {
+
+    private final BoundedElasticsearchSource source;
+
+    private RestClient restClient;
+    private String current;
+    private String scroll;
+
+    public BoundedElasticsearchReader(BoundedElasticsearchSource source) {
+      this.source = source;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      restClient = source.spec.getConnectionConfiguration().createClient();
+
+      String query = source.spec.getQuery();
+      if (query == null) {
+        query = "{\n"
+            + "  \"query\": {\n"
+            + "    \"match_all\": {}\n"
+            + "  }\n"
+            + "}";
+      }
+
+      Response response;
+      String endPoint = String.format("/%s/%s/_search",
+                                      source.spec.getConnectionConfiguration().getIndex(),
+                                      source.spec.getConnectionConfiguration().getType());
+      Map<String, String> params = new HashMap<>();
+      params.put("scroll", source.spec.getScrollKeepalive() != null ? source.spec
+          .getScrollKeepalive() : "1m");
+      params.put("size", "1");
+      if (source.shardPreference != null) {
+        params.put("preference", "_shards:" + source.shardPreference);
+      }
+      HttpEntity queryEntity = new NStringEntity(query, ContentType.APPLICATION_JSON);
+      response =
+          restClient.performRequest("GET", endPoint, params, queryEntity, new BasicHeader("", ""));
+      JsonObject searchResult = parseResponse(response);
+      scroll = searchResult.getAsJsonPrimitive("_scroll_id").getAsString();
+      return updateCurrent(searchResult);
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      String requestBody = String.format("{\"scroll\" : \"%s\",\"scroll_id\" : \"%s\"}",
+                                         source.spec.getScrollKeepalive() != null ? source.spec
```

Comment:
Hm, I don't quite like that this snippet (defaulting to 1m) is repeating. Can you structure the code in a way that getScrollKeepalive() is simply "1m" by default?